app/src/main/java/io/homeassistant/companion/android/widgets/WidgetPrefUtils.kt

```diff
@@ -0,0 +1,37 @@
+package io.homeassistant.companion.android.widgets
+
+import android.content.Context
+
+internal const val PREFS_NAME = "io.homeassistant.companion.android.widgets"
+internal const val PREF_PREFIX_KEY = "widget_"
+internal const val PREF_KEY_DOMAIN = "domain"
+internal const val PREF_KEY_SERVICE = "service"
+internal const val PREF_KEY_SERVICE_DATA = "serviceData"
+internal const val PREF_KEY_LABEL = "label"
+
+internal fun saveServiceCallData(
+    context: Context,
+    appWidgetId: Int,
+    domainStr: String,
+    serviceStr: String,
+    serviceDataStr: String
+) {
+    saveStringPref(context, appWidgetId, PREF_KEY_DOMAIN, domainStr)
+    saveStringPref(context, appWidgetId, PREF_KEY_SERVICE, serviceStr)
+    saveStringPref(context, appWidgetId, PREF_KEY_SERVICE_DATA, serviceDataStr)
+}
+
+internal fun saveStringPref(context: Context, appWidgetId: Int, key: String, data: String?) {
+    context.getSharedPreferences(PREFS_NAME, 0).edit()
+        .putString(PREF_PREFIX_KEY + appWidgetId + key, data)
+        .apply()
+}
+
+internal fun loadStringPref(context: Context, appWidgetId: Int, key: String): String? {
+    return context.getSharedPreferences(PREFS_NAME, 0)
+        .getString(PREF_PREFIX_KEY + appWidgetId + key, null)
+}
+
+internal fun deleteStringPref(context: Context, appWidgetId: Int, key: String) {
+    saveStringPref(context, appWidgetId, key, null)
+}
```

Comment:
These should be moved into a usecase and repository

Reply:
I'm an amateur developer (hence why I didn't write any tests) and I'm not sure how to properly implement this style.  Do you have an article or paper you would recommend I read to make sure I do this correctly?  Even just a style name would be helpful - the best I can find is "Clean Architecture" when doing a cursory search but I'm not sure that's correct.

Reply:
That's a good place to start, I would look at the classes you added the ```callService``` method too.  You would want something similar to that.

Reply:
Okay, I'll work on getting that set up.  I wasn't incredibly confident in how I was modifying the classes for `callService`, but if they look fine to you then I should be able to copy them well enough for a repository/implementation/usecase setup.

Reply:
This is complete in commit f877070